File: app/src/main/java/mx/dev/franco/automusictagfixer/UI/track_detail/TrackDetailViewModel.java
package mx.dev.franco.automusictagfixer.UI.track_detail;

import android.app.Application;
import android.arch.core.util.Function;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MediatorLiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.content.Context;
import android.support.annotation.NonNull;

import javax.inject.Inject;

import mx.dev.franco.automusictagfixer.R;
import mx.dev.franco.automusictagfixer.UI.SingleLiveEvent;
import mx.dev.franco.automusictagfixer.fixer.AudioMetadataTagger;
import mx.dev.franco.automusictagfixer.fixer.AudioTagger;
import mx.dev.franco.automusictagfixer.fixer.MetadataReaderResult;
import mx.dev.franco.automusictagfixer.fixer.MetadataWriterResult;
import mx.dev.franco.automusictagfixer.identifier.IdentificationParams;
import mx.dev.franco.automusictagfixer.persistence.repository.DataTrackRepository;
import mx.dev.franco.automusictagfixer.persistence.repository.IdentificationManager;
import mx.dev.franco.automusictagfixer.persistence.room.Track;
import mx.dev.franco.automusictagfixer.utilities.ActionableMessage;
import mx.dev.franco.automusictagfixer.utilities.AndroidUtils;
import mx.dev.franco.automusictagfixer.utilities.Constants;
import mx.dev.franco.automusictagfixer.utilities.IdentificationType;
import mx.dev.franco.automusictagfixer.utilities.Message;
import mx.dev.franco.automusictagfixer.utilities.Resource;

import static mx.dev.franco.automusictagfixer.common.Action.NONE;
import static mx.dev.franco.automusictagfixer.common.Action.SUCCESS_IDENTIFICATION;

public class TrackDetailViewModel extends AndroidViewModel {

    //Two-way bind livedata objects.
    public MutableLiveData<String> title;
    public MutableLiveData<String> artist;
    public MutableLiveData<String> album;
    public MutableLiveData<String> number;
    public MutableLiveData<String> year;
    public MutableLiveData<String> genre;
    public MutableLiveData<byte[]> cover;

    //One way bind livedata objects.
    public MutableLiveData<String> filesize;
    public MutableLiveData<String> channels;
    public MutableLiveData<String> type;
    public MutableLiveData<String> resolution;
    public MutableLiveData<String> frequency;
    public MutableLiveData<String> bitrate;
    public MutableLiveData<String> length;
    public MutableLiveData<String> absolutePath;

    //MediatorLiveData to observe loading state of multiple sources.
    private MediatorLiveData<Boolean> mStateMerger;

    private SingleLiveEvent<ValidationWrapper> mInputsInvalidLiveData = new SingleLiveEvent<>();
    private SingleLiveEvent<Message> mLiveMessage = new SingleLiveEvent<>();
    private SingleLiveEvent<ActionableMessage> mLiveActionableMessage = new SingleLiveEvent<>();
    private SingleLiveEvent<Void> mLiveConfirmationDeleteCover = new SingleLiveEvent<>();

    private LiveData<Message> mResultReading;
    private LiveData<ActionableMessage> mResultWriting;
    private LiveData<Message> mResultRenaming;
    private LiveData<IdentificationType> mResultsIdentificationLiveData;

    private DataTrackRepository mDataTrackRepository;
    private IdentificationManager mIdentificationManager;

    private AudioTagger.AudioFields mAudioFields;

    private int mCorrectionMode = Constants.CorrectionActions.VIEW_INFO;
    private ManualCorrectionParams mCorrectionParams;
    private IdentificationParams mIdentificationParams;

    @Inject
    public TrackDetailViewModel(@NonNull Application application,
                                @NonNull DataTrackRepository dataTrackRepository,
                                @NonNull IdentificationManager identificationManager) {
        super(application);
        mDataTrackRepository = dataTrackRepository;
        mIdentificationManager = identificationManager;

        title = new MutableLiveData<>();
        artist = new MutableLiveData<>();
        album = new MutableLiveData<>();
        number = new MutableLiveData<>();
        year = new MutableLiveData<>();
        genre = new MutableLiveData<>();
        cover = new MutableLiveData<>();

        filesize = new MutableLiveData<>();
        channels = new MutableLiveData<>();
        type = new MutableLiveData<>();
        resolution = new MutableLiveData<>();
        frequency = new MutableLiveData<>();
        bitrate = new MutableLiveData<>();
        length = new MutableLiveData<>();
        absolutePath = new MutableLiveData<>();

        mStateMerger = new MediatorLiveData<>();

        //Merge state loading into one live data to observe.
        LiveData<Boolean> stateTrackDataRepository = mDataTrackRepository.observeLoadingState();
        LiveData<Boolean> identificationRepositoryState = mIdentificationManager.observeLoadingState();
        mStateMerger.addSource(stateTrackDataRepository, aBoolean ->
                mStateMerger.setValue(aBoolean));

        mStateMerger.addSource(identificationRepositoryState, aBoolean ->
                mStateMerger.setValue(aBoolean));
    }

    /**
     * Livedata to inform to UI the progress of a task.
     * @return A live data object holding a boolean value.
     */
    public LiveData<Boolean> observeLoadingState() {
        return mStateMerger;
    }

    /**
     * Livedata to inform if track could be loaded.
     * @return Livedata holding a {@link Message} object or null.
     */
    public LiveData<Message> observeReadingResult() {
        LiveData<Resource<MetadataReaderResult>> resultReader = mDataTrackRepository.getResultReader();
        mResultReading = Transformations.map(resultReader, input -> {
            Message message = null;
            if(input.data.getFields().getCode() != AudioTagger.SUCCESS) {
                message = new Message(R.string.cannot_read_file);
            }
            else {
                mAudioFields = input.data.getFields();
                setEditableInfo(mAudioFields);
                setNoEditableInfo(mAudioFields);
                setFixedInfo(mAudioFields);
            }
            return message;
        });
        return mResultReading;
    }

    public LiveData<ActionableMessage> observeWritingResult() {
        LiveData<Resource<MetadataWriterResult>> resultWriter = mDataTrackRepository.getResultWriter();
        mResultWriting = Transformations.map(resultWriter, new Function<Resource<MetadataWriterResult>, ActionableMessage>() {
            @Override
            public ActionableMessage apply(Resource<MetadataWriterResult> input) {
                ActionableMessage message = null;
                if(input.data.getResultCorrection().getCode() == AudioTagger.SUCCESS) {
                    if(mCorrectionParams.renameFile()){
                        mDataTrackRepository.renameFile(mCorrectionParams);
                    }
                    else {
                        message = new ActionableMessage(NONE, R.string.changes_applied);
                    }
                }
                else {
                    message = new ActionableMessage(NONE,R.string.could_not_correct_file,
                            input.data.getResultCorrection().getError().getMessage());
                }
                return message;
            }
        });
        return mResultWriting;
    }

    public LiveData<Message> observeRenamingResult() {
        LiveData<Resource<AudioTagger.ResultRename>> resultRename = mDataTrackRepository.getResultRename();
        mResultRenaming = Transformations.map(resultRename, input -> {
            Message message;
            if(input.status == Resource.Status.SUCCESS) {
                message = new Message(R.string.changes_applied);
            }
            else {
                message = new Message(R.string.changes_partially_applied);
            }
            return message;
        });
        return mResultRenaming;
    }

    /**
     * Livedata to observe for identification tasks.
     * @return a Livedata holding the result.
     */
    public LiveData<IdentificationType> observeResultIdentification() {
        LiveData<Resource<ActionableMessage>> resultIdentification = mIdentificationManager.observeActionableMessage();
        mResultsIdentificationLiveData = Transformations.map(resultIdentification, input -> {

            if(input.data.getAction() == SUCCESS_IDENTIFICATION) {
                IdentificationType identificationType = new IdentificationType();
                identificationType.setIdentificationType(mIdentificationParams.getIdentificationType());
                return identificationType;
            }
            else {
                if(input.status == Resource.Status.ERROR || input.status == Resource.Status.CANCELLED)  {
                    mLiveActionableMessage.setValue(input.data);
                }
            }

            return null;
        });
        return mResultsIdentificationLiveData;
    }

    public LiveData<Message> observeMessage(){
        return mLiveMessage;
    }

    public LiveData<ActionableMessage> observeActionableMessage() {
        return mLiveActionableMessage;
    }


    public LiveData<ValidationWrapper> observeInvalidInputsValidation() {
        return mInputsInvalidLiveData;
    }

    public LiveData<Void> observeConfirmationRemoveCover() {
        return mLiveConfirmationDeleteCover;
    }

    public void setInitialAction(int correctionMode) {
        mCorrectionMode = correctionMode;
    }

    public void loadInfoTrack(int trackId) {
        mDataTrackRepository.loadDataTrack(trackId);
    }

    public void validateImageSize(ImageWrapper imageWrapper) {
        if (imageWrapper.height <= 2000 || imageWrapper.width <= 2000) {
            performCorrection();

        } else {
            Message message = new Message(R.string.image_too_big);
            mLiveMessage.setValue(message);
        }
    }

    public void confirmRemoveCover() {
        mDataTrackRepository.removeCover();
    }

    public void cancelIdentification() {
        mIdentificationManager.cancelIdentification();
    }

    public void performCorrection(AudioMetadataTagger.InputParams correctionParams) {
        mCorrectionParams = (ManualCorrectionParams) correctionParams;
        if(mCorrectionParams.getCorrectionMode() == Constants.MANUAL) {
            performCorrection();
        }
        else {
            mDataTrackRepository.performCorrection(correctionParams);
        }
    }

    private void performCorrection(){
        boolean inputsValid = inputsValid();
        if(inputsValid) {
            String title = this.title.getValue();
            String artist = this.artist.getValue();
            String album = this.album.getValue();
            String trackYear = this.year.getValue();
            String trackNumber = this.number.getValue();
            String genre = this.genre.getValue();

            if(title != null && !title.isEmpty())
                title = AudioTagger.StringUtilities.trimString(title);
            if(artist != null && !artist.isEmpty())
                artist = AudioTagger.StringUtilities.trimString(artist);
            if(album != null && !album.isEmpty())
                album = AudioTagger.StringUtilities.trimString(album);
            if(trackYear != null && !trackYear.isEmpty())
                trackYear = AudioTagger.StringUtilities.trimString(trackYear);
            if(trackNumber != null && !trackNumber.isEmpty())
                trackNumber = AudioTagger.StringUtilities.trimString(trackNumber);
            if(genre != null && !genre.isEmpty())
                genre = AudioTagger.StringUtilities.trimString(genre);
            if(this.)
            byte[] cover = this.cover.getValue();



            AudioMetadataTagger.InputParams inputParams = AndroidUtils.createInputParams(title,
                    artist,album, genre, trackNumber, trackYear, cover);
            inputParams.setTargetFile(getTrack().getPath());
            inputParams.setCodeRequest(AudioTagger.MODE_OVERWRITE_ALL_TAGS);
            mDataTrackRepository.performCorrection(inputParams);
        }

    }

    public void saveAsImageFileFrom(int cached) {

    }

    public void enableEditMode() {

    }

    public void restorePreviousValues() {

    }

    /**
     * Identifies the track.
     * @param identificationParams Input parameters that could be useful for this identification.
     */
    public void startIdentification(IdentificationParams identificationParams) {
        mIdentificationParams = identificationParams;
        mIdentificationManager.startIdentification(mDataTrackRepository.getTrack());
    }

    public void validateInputData() {

    }

    public void openInExternalApp(Context applicationContext) {

    }

    public void removeCover() {
        if(mAudioFields.getCover() != null) {

        }
        else {

        }
    }

    public Track getTrack() {
        return mDataTrackRepository.getTrack();
    }

    private void setNoEditableInfo(AudioTagger.AudioFields audioFields) {
        filesize.setValue(audioFields.getFileSize());
        absolutePath.setValue(audioFields.getPath());
    }

    private void setFixedInfo(AudioTagger.AudioFields audioFields) {
        channels.setValue(audioFields.getChannels());
        type.setValue(audioFields.getMimeType());
        resolution.setValue(audioFields.getResolution());
        frequency.setValue(audioFields.getFrequency());
        bitrate.setValue(audioFields.getBitrate());
        length.setValue(audioFields.getDuration());
    }

    private void setEditableInfo(AudioTagger.AudioFields audioFields) {
        title.setValue(audioFields.getTitle());
        artist.setValue(audioFields.getArtist());
        album.setValue(audioFields.getAlbum());
        number.setValue(audioFields.getTrackNumber());
        year.setValue(audioFields.getTrackYear());
        genre.setValue(audioFields.getGenre());
        cover.setValue(audioFields.getCover());
    }

    private boolean inputsValid(){
        String title = this.title.getValue();
        String artist = this.artist.getValue();
        String album = this.album.getValue();
        String trackYear = this.year.getValue();
        String trackNumber = this.number.getValue();
        String genre = this.genre.getValue();

        int field = -1;
        if(title != null){
            field = R.id.track_name_details;
            if(AudioTagger.StringUtilities.isFieldEmpty(title)) {
                mInputsInvalidLiveData.setValue(new ValidationWrapper(field, new Message(R.string.empty_tag)));
                return false;
            }
            if(AndroidUtils.isTooLong(field, title)) {
                mInputsInvalidLiveData.setValue(new ValidationWrapper(field, new Message(R.string.tag_too_long)));
                return false;
            }
            title = AudioTagger.StringUtilities.trimString(title);
        }

        if(artist != null){
            field = R.id.artist_name_details;
            if(AudioTagger.StringUtilities.isFieldEmpty(artist)) {
                mInputsInvalidLiveData.setValue(new ValidationWrapper(field, new Message(R.string.empty_tag)));
                return false;
            }
            if(AndroidUtils.isTooLong(field, artist)) {
                mInputsInvalidLiveData.setValue(new ValidationWrapper(field, new Message(R.string.tag_too_long)));
                return false;
            }
            artist = AudioTagger.StringUtilities.trimString(artist);
        }

        if(album != null){
            field = R.id.album_name_details;
            if(AudioTagger.StringUtilities.isFieldEmpty(album)) {
                mInputsInvalidLiveData.setValue(new ValidationWrapper(field, new Message(R.string.empty_tag)));
                return false;
            }
            if(AndroidUtils.isTooLong(field, album)) {
                mInputsInvalidLiveData.setValue(new ValidationWrapper(field, new Message(R.string.tag_too_long)));
                return false;
            }
            album = AudioTagger.StringUtilities.trimString(album);
        }

        if(trackYear != null){
            field = R.id.track_year;
            if(AudioTagger.StringUtilities.isFieldEmpty(trackYear)) {
                mInputsInvalidLiveData.setValue(new ValidationWrapper(field, new Message(R.string.empty_tag)));
                return false;
            }
            if(AndroidUtils.isTooLong(field, trackYear)) {
                mInputsInvalidLiveData.setValue(new ValidationWrapper(field, new Message(R.string.tag_too_long)));
                return false;
            }
            trackYear = AudioTagger.StringUtilities.trimString(trackYear);
        }

        if(trackNumber != null){
            field = R.id.track_number;
            if(AudioTagger.StringUtilities.isFieldEmpty(trackNumber)) {
                mInputsInvalidLiveData.setValue(new ValidationWrapper(field, new Message(R.string.empty_tag)));
                return false;
            }
            if(AndroidUtils.isTooLong(field, trackNumber)) {
                mInputsInvalidLiveData.setValue(new ValidationWrapper(field, new Message(R.string.tag_too_long)));
                return false;
            }
            trackNumber = AudioTagger.StringUtilities.trimString(trackNumber);
        }

        if(genre != null){
            field = R.id.track_genre;
            if(AudioTagger.StringUtilities.isFieldEmpty(genre)) {
                mInputsInvalidLiveData.setValue(new ValidationWrapper(field, new Message(R.string.empty_tag)));
                return false;
            }
            if(AndroidUtils.isTooLong(field, genre)) {
                mInputsInvalidLiveData.setValue(new ValidationWrapper(field, new Message(R.string.tag_too_long)));
                return false;
            }
            genre = AudioTagger.StringUtilities.trimString(genre);
        }

        return true;
    }

    @Override
    protected void onCleared() {
        mDataTrackRepository.onCleared();
    }
}

File: app/src/main/java/mx/dev/franco/automusictagfixer/ui/trackdetail/TrackDetailViewModel.kt
package mx.dev.franco.automusictagfixer.ui.trackdetail

import android.app.Application
import android.util.ArrayMap
import android.util.Log
import androidx.lifecycle.*
import com.google.android.material.snackbar.Snackbar
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import mx.dev.franco.automusictagfixer.R
import mx.dev.franco.automusictagfixer.common.Action
import mx.dev.franco.automusictagfixer.covermanager.CoverLoader
import mx.dev.franco.automusictagfixer.filemanager.ImageFileSaver
import mx.dev.franco.automusictagfixer.fixer.AudioTagger
import mx.dev.franco.automusictagfixer.fixer.AudioTagger.*
import mx.dev.franco.automusictagfixer.fixer.CorrectionParams
import mx.dev.franco.automusictagfixer.persistence.mediastore.MediaStoreAccess
import mx.dev.franco.automusictagfixer.persistence.repository.TrackManager
import mx.dev.franco.automusictagfixer.persistence.room.Track
import mx.dev.franco.automusictagfixer.ui.SingleLiveEvent
import mx.dev.franco.automusictagfixer.utilities.AndroidUtils
import mx.dev.franco.automusictagfixer.utilities.Constants
import mx.dev.franco.automusictagfixer.utilities.SnackbarMessage
import mx.dev.franco.automusictagfixer.utilities.resource_manager.ResourceManager
import org.jaudiotagger.audio.exceptions.CannotReadException
import org.jaudiotagger.audio.exceptions.InvalidAudioFrameException
import org.jaudiotagger.audio.exceptions.ReadOnlyFileException
import org.jaudiotagger.tag.FieldKey
import org.jaudiotagger.tag.TagException
import java.io.IOException
import javax.inject.Inject
import kotlin.collections.set


class TrackDetailViewModel
@Inject
constructor(
    private val trackManager: TrackManager,
    private val mediaStoreAccess: MediaStoreAccess,
    private val audioTagger: AudioTagger,
    private val resourceManager: ResourceManager,
    application: Application
) : AndroidViewModel(application) {
    private val mTrackLoader = MediatorLiveData<Track>()

    //Two-way bind livedata objects.

    var title = MutableLiveData<String?>()

    var artist = MutableLiveData<String>()

    var album = MutableLiveData<String>()

    var number = MutableLiveData<String>()

    var year = MutableLiveData<String>()

    var genre = MutableLiveData<String>()
    val cover: MutableLiveData<ByteArray?> by lazy { MutableLiveData<ByteArray?>() }
    var imageSize = MutableLiveData<String>()
    var isStoredInSD = SingleLiveEvent<Boolean>()

    //One way bind livedata objects.

    var filesize = MutableLiveData<String>()

    var channels = MutableLiveData<String>()

    var type = MutableLiveData<String>()

    var resolution = MutableLiveData<String>()

    var frequency = MutableLiveData<String>()

    val bitrate = MutableLiveData<String>()

    val length = MutableLiveData<String>()

    val absolutePath = MutableLiveData<String>()
    val filename: MutableLiveData<String> by lazy { MutableLiveData<String>() }
    private val mLiveInformativeMessage: SingleLiveEvent<Int> = SingleLiveEvent()

    //MediatorLiveData to observe loading state of multiple sources.
    private val busy: MutableLiveData<Boolean> by lazy { MutableLiveData<Boolean>() }

    private val snackbarMessage: MutableLiveData<SnackbarMessage<String>> by lazy {
        MutableLiveData<SnackbarMessage<String>>()
    }

    private val mMessageMerger = MediatorLiveData<String>()
    private val mInputsInvalidLiveData = MutableLiveData<ValidationWrapper>()
    private val mLiveConfirmationDeleteCover = SingleLiveEvent<Void>()
    private val mResultReading = MediatorLiveData<AudioFields>()
    private val mLiveSuccessReading = MutableLiveData<Unit>()
    private val mResultWriting = MediatorLiveData<Map<FieldKey, Any>>()
    private var mAudioFields: AudioFields? = null
    private var mInitialAction = Constants.CorrectionActions.VIEW_INFO



    init {
        /*mResultWriting.addSource(this.trackManager.observeWritingResult()) { writingResult: AudioTaggerResult<Map<FieldKey, Any>> ->
            onWritingResult(
                writingResult
            )
        }*/
    }

    /**
     * Load the  information for current track.
     * @param trackId The id of track to load.
     */
    fun loadInfoTrack(trackId: Int) {
        busy.value = true
        viewModelScope.launch(Dispatchers.IO) {
            val track = trackManager.getTrack(trackId)
             val trackData = try {
                audioTagger.readFile(track.path) as AudioFields
            } catch (e: ReadOnlyFileException) {
                e.printStackTrace()
                AudioFields(COULD_NOT_READ_TAGS, e)
            } catch (e: CannotReadException) {
                e.printStackTrace()
                AudioFields(COULD_NOT_READ_TAGS, e)
            } catch (e: TagException) {
                e.printStackTrace()
                AudioFields(COULD_NOT_READ_TAGS, e)
            } catch (e: InvalidAudioFrameException) {
                e.printStackTrace()
                AudioFields(COULD_NOT_READ_TAGS, e)
            } catch (e: IOException) {
                e.printStackTrace()
                AudioFields(COULD_NOT_READ_TAGS, e)
            }

            withContext(Dispatchers.Main) {
                onReadingResult(trackData)
                busy.value = false
            }
        }
    }

    /**
     * Livedata to inform to UI the progress of a task.
     * @return A live data object holding a boolean value.
     */
    fun observeLoadingState() = busy as LiveData<Boolean>

    fun observeReadingResult(): LiveData<Unit> {
        return mLiveSuccessReading
    }

    fun observeAudioData(): LiveData<Void> {
        return Transformations.map(mResultReading) { input: AudioFields? -> null }
    }

    fun observeWritingFinishedEvent(): LiveData<Void> {
        return Transformations.map(mResultWriting) { input: Map<FieldKey, Any>? -> null }
    }

    fun observeInvalidInputsValidation(): LiveData<ValidationWrapper> {
        return mInputsInvalidLiveData
    }

    fun onMessage(): LiveData<String> {
        return mMessageMerger
    }

    fun observeConfirmationRemoveCover(): LiveData<Void> {
        return mLiveConfirmationDeleteCover as LiveData<Void>
    }

    fun observeCoverSavingResult() = snackbarMessage as LiveData<SnackbarMessage<String>>


    fun observeLoadingMessage()= mLiveInformativeMessage as LiveData<Int>

    fun observeIsStoredInSD() = isStoredInSD as LiveData<Boolean>

    private fun onReadingResult(audioFields: AudioFields) {
        if (audioFields.code != AudioTagger.SUCCESS) {
            mLiveInformativeMessage.value = R.string.could_not_read_file
            isStoredInSD.setValue(false)
        } else {
            mAudioFields = audioFields
            setEditableInfo(audioFields)
            setNoEditableInfo(audioFields)
            setFixedInfo(audioFields)
            Log.e(javaClass.name, "mLiveSuccessReading.setValue")
            mLiveSuccessReading.value = Unit
            if (audioFields.isStoredInSD) {
                isStoredInSD.setValue(true)
            } else {
                isStoredInSD.setValue(false)
            }
        }
    }

    /*fun observeErrorWriting(): LiveData<SnackbarMessage<*>> {
        return Transformations.map(mTrackManager.observeErrorWriting()) { input: AudioTaggerResult<Map<FieldKey, Any>> ->
            getMessageErrorWriting(
                input
            )
        }
    }*/

    /*private fun getMessageErrorWriting(input: AudioTaggerResult<Map<FieldKey, Any>>): SnackbarMessage<*> {
        val builder: SnackbarMessage.Builder<*> = SnackbarMessage.Builder<Any?>(
            getApplication()
        )
        val errorCode = input.code
        if (errorCode == AudioTagger.COULD_NOT_GET_URI_SD_ROOT_TREE) {
            builder.action(Action.URI_ERROR).body(R.string.message_uri_tree_not_set)
                .mainActionText(R.string.details).dismissible(false)
        } else {
            val idStringResource: Int
            idStringResource = when (errorCode) {
                AudioTagger.COULD_NOT_REMOVE_COVER, AudioTagger.COULD_NOT_APPLY_COVER -> R.string.message_could_not_apply_cover
                AudioTagger.COULD_NOT_COPY_BACK_TO_ORIGINAL_LOCATION, AudioTagger.COULD_NOT_CREATE_AUDIOFILE, AudioTagger.COULD_NOT_CREATE_TEMP_FILE, AudioTagger.COULD_NOT_REMOVE_OLD_ID3_VERSION, AudioTagger.COULD_NOT_RENAME_FILE, AudioTagger.COULD_NOT_WRITE_TAGS, AudioTagger.COULD_NOT_APPLY_TAGS -> R.string.message_could_not_apply_tags
                else -> R.string.message_could_not_apply_tags
            }
            builder.action(Action.NONE)
            builder.body(idStringResource)
        }
        return builder.build()
    }*/

    private fun onWritingResult(writingResult: AudioTaggerResult<Map<FieldKey, Any>>) {
        mResultWriting.value = writingResult.data
        mLiveInformativeMessage.value = R.string.changes_applied
        val deleteCoverFromCache =
            writingResult.taskExecuted == AudioTagger.MODE_OVERWRITE_ALL_TAGS ||
                    writingResult.taskExecuted == MODE_ADD_COVER ||
                    writingResult.taskExecuted == MODE_REMOVE_COVER
        if (deleteCoverFromCache) CoverLoader.removeCover(currentTrack.mediaStoreId.toString() + "")
    }

    /**
     * Change the cover without asking to user once a valid image is picked up.
     * @param imageWrapper
     */
    fun fastCoverChange(imageWrapper: ImageWrapper) {
        if (imageWrapper.height <= ImageWrapper.MAX_HEIGHT && imageWrapper.width <= ImageWrapper.MAX_WIDTH) {
            val correctionParams = CorrectionParams()
            correctionParams.correctionMode = MODE_ADD_COVER
            correctionParams.target = trackManager.currentTrack!!.path
            correctionParams.setRenameFile(false)
            val tags: MutableMap<FieldKey, Any> = ArrayMap()
            viewModelScope.launch(Dispatchers.IO) {
                busy.postValue(true)
                val data = AndroidUtils.generateCover(imageWrapper.bitmap)
                tags[FieldKey.COVER_ART] = data
                busy.postValue(false)
                correctionParams.tags = tags
                if (imageWrapper.requestCode == TrackDetailFragment.INTENT_GET_AND_UPDATE_FROM_GALLERY) {
                    val audioTaggerResult = trackManager.performCorrection(correctionParams)
                } else {
                    cover.postValue(data)
                }
            }
        } else {
            mLiveInformativeMessage.setValue(R.string.image_too_big)
        }
    }

    /**
     * Remove the cover.
     */
    fun confirmRemoveCover() {
        if (mAudioFields!!.cover != null) {
            busy.value = true
            mLiveInformativeMessage.value = R.string.removing_cover
            val correctionParams = CorrectionParams()
            correctionParams.target = currentTrack.path
            correctionParams.correctionMode = MODE_REMOVE_COVER
            //trackManager.performCorrection(correctionParams)
        } else {
            mLiveInformativeMessage.setValue(R.string.does_not_exist_cover)
        }
    }

    /**
     * Executes the correction of current track.
     * @param correctionParams The params required to correct current track.
     */
    fun performCorrection(correctionParams: CorrectionParams) {
        busy.value = true
        correctionParams.target = currentTrack.path
        if (correctionParams.tagsSource == Constants.MANUAL) {
            performManualCorrection(correctionParams)
        } else {
            val codeRequest = correctionParams.correctionMode
            mLiveInformativeMessage.setValue(R.string.applying_tags)
            when (codeRequest) {
                MODE_ADD_COVER -> performAddCoverCorrection(correctionParams)
                AudioTagger.MODE_OVERWRITE_ALL_TAGS, AudioTagger.MODE_WRITE_ONLY_MISSING -> {
                }
            }
        }
    }

    fun renameFile(correctionParams: CorrectionParams) {
        correctionParams.target = currentTrack.path
        //trackManager.performCorrection(correctionParams)
    }

    fun saveAsImageFileFrom(id: String?) {
        val newFilename = AndroidUtils.generateNameWithDate(mAudioFields!!.fileName)
        mLiveInformativeMessage.setValue(R.string.saving_cover)
    }

    fun extractCover() {
        val cover = cover.value
        if (cover != null) {
            var imageName: String? = null
            imageName = if (title.value != null && !title.value!!.isEmpty()) {
                StringUtilities.sanitizeString(title.value).trim { it <= ' ' }.replace(" ", "_")
            } else {
                ImageFileSaver.GENERIC_NAME
            }
            mLiveInformativeMessage.value = R.string.saving_cover
            val builder: SnackbarMessage.Builder<String> =
                SnackbarMessage.Builder(getApplication())

            viewModelScope.launch(Dispatchers.IO) {
                val s = ImageFileSaver.saveImageFile(cover, AndroidUtils.generateNameWithDate(imageName))

                withContext(Dispatchers.Main) {

                    if(ImageFileSaver.INPUT_OUTPUT_ERROR == s || ImageFileSaver.NULL_DATA == s
                        || ImageFileSaver.NO_EXTERNAL_STORAGE_WRITABLE == s
                    ) {
                        builder.body(R.string.cover_not_saved);
                        mediaStoreAccess.addFileToMediaStore(s, null);
                    }
                    else {
                        builder
                            .action(Action.WATCH_IMAGE).mainActionText(R.string.see_image)
                            .duration(Snackbar.LENGTH_LONG)
                            .data(s)
                            .body(R.string.cover_saved)
                    }
                    snackbarMessage.value = builder.build()
                }
            }
        } else {
            mLiveInformativeMessage.setValue(R.string.does_not_exist_cover)
        }
    }

    val currentTrack: Track
        get() = trackManager.currentTrack!!

    fun setInitialAction(action: Int) {
        mInitialAction = action
    }

    fun restorePreviousValues() {
        setEditableInfo(mAudioFields)
    }

    fun removeCover() {
        if (cover.value != null) {
            mLiveConfirmationDeleteCover.setValue(null)
        } else {
            mLiveInformativeMessage.setValue(R.string.does_not_exist_cover)
        }
    }

    private fun performAddCoverCorrection(correctionParam: CorrectionParams) {
        //Result result = findResult(correctionParam.getCoverId());
        //applyCoverFromRemote(result, correctionParam);
    }

    /*private void applyCoverFromRemote(Result result, CorrectionParams correctionParam) {
        Thread thread = new Thread(() -> {
            try {
                boolean hasInternet = AndroidUtils.isConnected(getApplication());

                if (!hasInternet) {
                    mLoadingStateMerger.postValue(false);
                    mLiveInformativeMessage.postValue(R.string.connect_to_internet);
                }
                else {
                    final byte[] coverData = GnUtils.fetchGnCover(result.getCoverArt().getUrl(),
                            getApplication());

                    if (coverData == null) {
                        mLoadingStateMerger.postValue(false);
                        mLiveInformativeMessage.postValue(R.string.error_fetching_image);
                    }
                    else {
                        mHandler.post(() -> {
                            Map<FieldKey, Object> tags = new ArrayMap<>();
                            tags.put(FieldKey.COVER_ART, coverData);
                            correctionParam.setTags(tags);
                            mTrackManager.performCorrection(correctionParam);
                        });
                    }
                }
            } catch (GnException e) {
                e.printStackTrace();
                mLoadingStateMerger.setValue(false);
                mLiveInformativeMessage.postValue(R.string.error_fetching_image);
            }
        });
        thread.start();
    }*/
    /*private void performSemiautomaticCorrection(CorrectionParams correctionParams) {
        mLoadingStateMerger.setValue(true);
        final Result result = findResult(correctionParams.getTrackId());
        String url = result.getCoverArt() != null ? result.getCoverArt().getUrl() : null;
        Thread thread = new Thread(() -> {
            byte[] data = null;
            if (url != null) {
                try {
                    data = GnUtils.fetchGnCover(url, getApplication());
                } catch (GnException e) {
                    e.printStackTrace();
                }
            }
            AndroidUtils.createInputParams(result.getTitle(),
                    result.getArtist(),
                    result.getAlbum(),
                    result.getGenre(),
                    result.getTrackNumber(),
                    result.getTrackYear(),
                    data,
                    correctionParams);

            mHandler.post(() -> mTrackManager.performCorrection(correctionParams));
        });
        thread.start();

    }*/
    private fun performManualCorrection(correctionParams: CorrectionParams) {
        val validationWrapper = inputsValid()
        if (validationWrapper == null) {
            var title = title.value
            var artist = artist.value
            var album = album.value
            var trackYear = year.value
            var trackNumber = number.value
            var genre = genre.value
            val cover = cover.value
            if (title != null && !title.isEmpty()) title = StringUtilities.trimString(title)
            if (artist != null && !artist.isEmpty()) artist = StringUtilities.trimString(artist)
            if (album != null && !album.isEmpty()) album = StringUtilities.trimString(album)
            if (trackYear != null && !trackYear.isEmpty()) trackYear =
                StringUtilities.trimString(trackYear)
            if (trackNumber != null && !trackNumber.isEmpty()) trackNumber =
                StringUtilities.trimString(trackNumber)
            if (genre != null && !genre.isEmpty()) genre = StringUtilities.trimString(genre)
            AndroidUtils.createInputParams(
                title,
                artist, album, genre, trackNumber, trackYear, cover, correctionParams
            )
            correctionParams.correctionMode = AudioTagger.MODE_OVERWRITE_ALL_TAGS
            mLiveInformativeMessage.setValue(R.string.applying_tags)
            //trackManager.performCorrection(correctionParams)
        } else {
            busy.setValue(false)
            validationWrapper.message = R.string.empty_tag
            mInputsInvalidLiveData.setValue(validationWrapper)
        }
    }

    private fun setNoEditableInfo(audioFields: AudioFields) {
        filesize.value = audioFields.fileSize
        absolutePath.value = audioFields.path + "/" + audioFields.fileName
        val coverSize = if (audioFields.imageSize != null)
            audioFields.imageSize + " " + resourceManager.getString(R.string.pixels)
        else
            resourceManager.getString(R.string.missing_cover)
        imageSize.value = coverSize
        filename.value = audioFields.fileName
    }

    private fun setFixedInfo(audioFields: AudioFields) {
        channels.value = audioFields.channels
        type.value = audioFields.mimeType
        resolution.value = audioFields.resolution
        frequency.value = audioFields.frequency
        bitrate.value = audioFields.bitrate
        length.value = audioFields.duration
    }

    private fun setEditableInfo(audioFields: AudioFields?) {
        title.value = audioFields!!.title
        artist.value = audioFields.artist
        album.value = audioFields.album
        number.value = audioFields.trackNumber
        year.value = audioFields.trackYear
        genre.value = audioFields.genre
        cover.value = audioFields.cover
    }

    private fun inputsValid(): ValidationWrapper? {
        val title = title.value
        val artist = artist.value
        val album = album.value
        val trackYear = year.value
        val trackNumber = number.value
        val genre = genre.value
        var field = -1
        if (title != null) {
            field = R.id.track_name_details
            if (StringUtilities.isFieldEmpty(title)) {
                return ValidationWrapper(field, R.string.empty_tag)
            }
            if (AndroidUtils.isTooLong(field, title)) {
                return ValidationWrapper(field, R.string.tag_too_long)
            }
        }
        if (artist != null) {
            field = R.id.artist_name_details
            if (StringUtilities.isFieldEmpty(artist)) {
                return ValidationWrapper(field, R.string.empty_tag)
            }
            if (AndroidUtils.isTooLong(field, artist)) {
                return ValidationWrapper(field, R.string.tag_too_long)
            }
        }
        if (album != null) {
            field = R.id.album_name_details
            if (StringUtilities.isFieldEmpty(album)) {
                return ValidationWrapper(field, R.string.empty_tag)
            }
            if (AndroidUtils.isTooLong(field, album)) {
                return ValidationWrapper(field, R.string.tag_too_long)
            }
        }
        if (trackYear != null) {
            field = R.id.track_year
            if (StringUtilities.isFieldEmpty(trackYear)) {
                return ValidationWrapper(field, R.string.empty_tag)
            }
            if (AndroidUtils.isTooLong(field, trackYear)) {
                return ValidationWrapper(field, R.string.tag_too_long)
            }
        }
        if (trackNumber != null) {
            field = R.id.track_number
            if (StringUtilities.isFieldEmpty(trackNumber)) {
                return ValidationWrapper(field, R.string.empty_tag)
            }
            if (AndroidUtils.isTooLong(field, trackNumber)) {
                return ValidationWrapper(field, R.string.tag_too_long)
            }
        }
        if (genre != null) {
            field = R.id.track_genre
            if (StringUtilities.isFieldEmpty(genre)) {
                return ValidationWrapper(field, R.string.empty_tag)
            }
            if (AndroidUtils.isTooLong(field, genre)) {
                return ValidationWrapper(field, R.string.tag_too_long)
            }
        }
        return null
    }
}